Added from_scipy_sparse_matrix function

salfaris opened the pull request:
I've added a simple function (together with tests) to convert a scipy sparse matrix into a jraph GraphsTuple as per mentioned in issue #36. Let me know if anything need changes :)

Some future work I'm considering includes making a to_scipy_sparse_matrix function and intermediate functions for NetworkX.

jg8610 commented on jraph/_src/utils.py:
```diff
@@ -1122,3 +1123,33 @@ def wrapper(graph: gn_graph.GraphsTuple) -> gn_graph.GraphsTuple:
   return wrapper
+
+
+def from_scipy_sparse_matrix(A: scipy.sparse.spmatrix):
```
Could you rename to sparse_matrix_to_graphs_tuple

jg8610 commented on jraph/_src/utils.py:
```diff
@@ -1122,3 +1123,33 @@ def wrapper(graph: gn_graph.GraphsTuple) -> gn_graph.GraphsTuple:
+  """Creates a `jraph.GraphsTuple` from a scipy sparse matrix.
+  
+  Args:
+    A: A SciPy sparse matrix.
```
could you rename this variable to sparse_matrix

salfaris commented on jraph/_src/utils.py:
```diff
@@ -1122,3 +1123,33 @@ def wrapper(graph: gn_graph.GraphsTuple) -> gn_graph.GraphsTuple:
   return wrapper
+
+
+def from_scipy_sparse_matrix(A: scipy.sparse.spmatrix):
```
Yeah fair point, let me see what I can do

jg8610 replied:
Thanks for your patience while we think through this.

We're not sure this is general enough to go in the jraph utilities, but we can see the value of having this available. One option may to add it as an example, or add it to documentation, so that we can remove dependency requirements from the main library.

What do you think?

salfaris replied:
I don't know why I did not think about this before. I think we can get away without importing `scipy` at all. We just have to be more explicit on which sparse matrix format we use. Fixing the COO sparse format should be enough where we expect users to pass in `senders`, `receivers`, `value` (for the number of edges between the sender and receiver) and `n_node`.

What do you think about this? Otherwise, I would be happy to add it as an example/documentation :)

jg8610 commented on jraph/_src/utils.py:
```diff
@@ -1122,3 +1123,33 @@ def wrapper(graph: gn_graph.GraphsTuple) -> gn_graph.GraphsTuple:
+  
+  senders = []
+  receivers = []
+  for idx, (sender_idx, receiver_idx) in enumerate(zip(A.row, A.col)):
```
I would prefer using np.repeat here, it will be a lot lot faster.

salfaris replied:
This suggestion is amazing, I didn't know we can do this!

jg8610 commented on jraph/_src/utils.py:
```diff
@@ -1122,3 +1123,33 @@ def wrapper(graph: gn_graph.GraphsTuple) -> gn_graph.GraphsTuple:
+  
+  return gn_graph.GraphsTuple(
+    nodes=None,
+    edges=None,
```
For utilities could we stick to np arrays

jg8610 commented on jraph/_src/utils.py:
```diff
@@ -1122,3 +1123,33 @@ def wrapper(graph: gn_graph.GraphsTuple) -> gn_graph.GraphsTuple:
+    A `jraph.GraphsTuple` graph based on the scipy sparse matrix `A`.
+  """
+  A = A.tocoo()
+  n_node = jnp.asarray([A.shape[1]])
```
could we stick to numpy arrays here and below.

salfaris replied:
I'm sorry but I'm a bit confused on where should we use numpy arrays and where should we use jax.numpy arrays in utilities? I've seen other code in utilities using a bit of both so I'm not entirely sure.

salfaris commented:
So I've updated `sparse_matrix_to_graphs_tuple` to be independent of scipy. Plus, I've implemented it using `np.repeat` as you suggested making it not only faster but cleaner as well. Let me know what you think about this @jg8610 :)

salfaris commented:
> So I've updated `sparse_matrix_to_graphs_tuple` to be independent of scipy. Plus, I've implemented it using `np.repeat` as you suggested making it not only faster but cleaner as well. Let me know what you think about this @jg8610 :)

Hey @jg8610, would be great if you can check this out thanks 😄

jg8610 commented:
Thanks for bearing with me! I think this looks good, if you could resolve the final conflict I'll accept it :)

salfaris commented:
No worries at all @jg8610 :) 

And conflict is resolved!

jg8610 commented on jraph/_src/utils_test.py:
```diff
@@ -1056,5 +1119,19 @@ def test_zero_padding_values(self, wrapper):
+    for (sparse_matrix,
+         expected_graph) in zip(sparse_adj_matrices, expected_graphs):
+      senders, receivers, values, n_node = sparse_matrix
+      from_sparse_graph = utils.\
```
could you avoid using a backslash here for line continuation, thanks!

jg8610 commented on jraph/_src/utils_test.py:
```diff
@@ -167,6 +167,69 @@ def _get_list_and_batched_graph():
 
 
+def _get_list_matrix():
+  """Returns a list of adjacency matrices, its sparse version
```
could you make this comment fit on one line?

Returns sparse and graphstuple version of adj matrix.

jg8610 commented on jraph/_src/utils_test.py:
```diff
@@ -167,6 +167,69 @@ def _get_list_and_batched_graph():
+      np.array([[0, 0], [1, 0]]),
+  ]
+  # Sparse version of the above adjacency matrix.
+  sparse_COO_matrices = [
```
could you remove capital letters in variable names,

jg8610 commented on jraph/_src/utils.py:
```diff
@@ -1122,3 +1122,32 @@ def wrapper(graph: gn_graph.GraphsTuple) -> gn_graph.GraphsTuple:
 
   return wrapper
+
+  
```
please leave only one line break

salfaris replied:
Done, but just curious when should you leave one line break and when should you leave two? Because in `utils.py`, I think all have two line breaks?

jg8610 replied:
Sorry, this one was my bad when copying over, this really is the last one, I should give you some sort of kudos for being so patient with me :)

salfaris replied:
No worries at all :)

So do you want me to make it two line breaks again?

jg8610 replied:
yes please, thank you!

salfaris replied:
Done and pushed!

jg8610 commented on jraph/_src/utils.py:
```diff
@@ -1122,3 +1122,32 @@ def wrapper(graph: gn_graph.GraphsTuple) -> gn_graph.GraphsTuple:
+    senders: jnp.ndarray, receivers: jnp.ndarray,
+    values: jnp.ndarray, n_node: jnp.ndarray) -> gn_graph.GraphsTuple:
+  """Creates a `jraph.GraphsTuple` from a sparse matrix in COO format.
+  
```
trailing line space, please delete

jg8610 replied:
*trailing white space

jg8610 commented on jraph/_src/utils.py:
```diff
@@ -1122,3 +1122,32 @@ def wrapper(graph: gn_graph.GraphsTuple) -> gn_graph.GraphsTuple:
+    receivers: The column indices of the matrix.
+    values: The values of the matrix.
+    n_node: The number of nodes in the graph defined by the sparse matrix.
+    
```
trailing white space, please delete

jg8610 reviewed (commented):
Thank you so much, I received these linting errors when trying to merge, please fix and then we'll be done!
